overlord/snapstate/handlers.go

```diff
@@ -1570,7 +1570,24 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 
 	// if we just installed a core snap, request a restart
 	// so that we switch executing its snapd.
-	m.maybeRestart(t, newInfo, reboot)
+	var canReboot bool
+	if reboot {
+		var cannotReboot bool
+		// system reboot is required, but can this task request that?
+		if err := t.Get("cannot-reboot", &cannotReboot); err != nil && err != state.ErrNoState {
+			return err
+		}
+		if !cannotReboot {
+			// either the task was created before that variable was
+			// introduced or the task can request a reboot
+			canReboot = true
+		} else {
+			t.Logf("reboot postponed to later tasks")
+		}
+	}
+	if !reboot || canReboot {
```

Comment:
Shouldn't this be `if reboot || canReboot {`?

I know that you didn't introduce the `reboot` variable with this change, but IMHO this branch would be a good occasion to rename it (`mustReboot`, `reboot required`, `willReboot`, `didReboot`,etc., depending on what the variable actually means).

Reply:
Hmm `needsReboot`?

As for the if check, I see how it can be confusing. I'll tweak this part to make it more clear, but yeah the condition is `!reboot || canReboot` (canReboot is only true when we need a reboot and it's allowed, and !reboot is true when there may be a need to daemon restart only)

Reply:
renamed to needsReboot